programs/cleos/do_http_post_libcurl.cpp

```diff
@@ -0,0 +1,135 @@
+
+
+#include "do_http_post.hpp"
+#include <curl/curl.h>
+#include <eosio/chain/exceptions.hpp>
+
+namespace eosio { namespace client { namespace http {
+
+   static size_t write_callback(void* data, size_t size, size_t nmemb, void* userp) {
+      size_t       realsize = size * nmemb;
+      std::string& mem      = *(std::string*)userp;
+      mem.append((const char*)data, realsize);
+      return realsize;
+   }
+
+   void dump(const char* text, FILE* stream, unsigned char* ptr, size_t size) {
+      size_t       i;
+      size_t       c;
+      unsigned int width = 0x10;
+
+      fprintf(stream, "%s, %10.10ld bytes (0x%8.8lx)\n", text, (long)size, (long)size);
+
+      for (i = 0; i < size; i += width) {
+         fprintf(stream, "%4.4lx: ", (long)i);
+
+         /* show hex to the left */
+         for (c = 0; c < width; c++) {
+            if (i + c < size)
+               fprintf(stream, "%02x ", ptr[i + c]);
+            else
+               fputs("   ", stream);
+         }
+
+         /* show data on the right */
+         for (c = 0; (c < width) && (i + c < size); c++) {
+            char x = (ptr[i + c] >= 0x20 && ptr[i + c] < 0x80) ? ptr[i + c] : '.';
+            fputc(x, stream);
+         }
+
+         fputc('\n', stream); /* newline */
+      }
+   }
+
+   int my_trace(CURL* handle, curl_infotype type, char* data, size_t size, void* userp) {
+      const char* text;
+      (void)handle; /* prevent compiler warning */
+      (void)userp;
+
+      switch (type) {
+         case CURLINFO_TEXT: fprintf(stderr, "== Info: %s", data);
+         default: /* in case a new one is introduced to shock us */ return 0;
+
+         case CURLINFO_HEADER_OUT: text = "=> Send header"; break;
+         case CURLINFO_DATA_OUT: text = "=> Send data"; break;
+         case CURLINFO_SSL_DATA_OUT: text = "=> Send SSL data"; break;
+         case CURLINFO_HEADER_IN: text = "<= Recv header"; break;
+         case CURLINFO_DATA_IN: text = "<= Recv data"; break;
+         case CURLINFO_SSL_DATA_IN: text = "<= Recv SSL data"; break;
+      }
+
+      dump(text, stderr, (unsigned char*)data, size);
+      return 0;
+   }
+
+   std::tuple<unsigned int, std::string> do_http_post(const std::string& base_uri, const std::string& path,
+                                                      const std::vector<std::string>& headers,
+                                                      const std::string& postjson, bool verify_cert, bool verbose,
+                                                      bool trace) {
+
+      static bool initialized = false;
+
+      if (!initialized) {
+         auto res = curl_global_init(CURL_GLOBAL_DEFAULT);
+         EOS_ASSERT(res == CURLE_OK, chain::http_exception, curl_easy_strerror(res));
+         initialized = true;
+      }
+
+      auto curl = curl_easy_init();
```

Comment:
I think it would be good to use a `fc::make_scoped_exit` to `curl_easy_cleanup`. There is nothing currently I see that would throw outside of `bad_alloc`, but it would possibly protect against code changes that might introduce exceptions.

Reply:
Does calling `curl_easy_cleanup()` in here prevent keep alive though?

Reply:
I believe so. However, in the origin asio implementation, it creates new connection per http request. There's no keep alive mechanism either. To make that happen, some bookkeeping mechanism to associates URL and the curl handle is needed. I am not sure it's worth the effort given the fact that this is just a CLI utility.

Reply:
I think it's as straightforward as just reusing the same curl handle and libcurl sorts it out itself -- even for different servers and whatnot. It's not unusual for cleos to make more than one request, and the roundtrip time of starting a new TLS connection is non-trivial.

But yeah, the previous code doesn't do it so improvements can be addressed later. I thought we added proper support for this in cleos at the same time as http_plugin (included in 3.2) but apparently not.